rules/onchain_budget.go

```diff
@@ -0,0 +1,613 @@
+	}
+
+	checker, ok := checkers[uri]
+	if !ok {
+		return nil, nil
+	}
+
+	if !checker.HandlesRequest(msg.ProtoReflect().Type()) {
+		return nil, fmt.Errorf("invalid implementation, checker "+
+			"for URI %s does not accept request of type %v",
+			uri, msg.ProtoReflect().Type())
+	}
+
+	return checker.HandleRequest(ctx, msg)
+}
+
+// HandleResponse handles and possible alters a response.
+//
+// NOTE: this is part of the Rule interface.
+func (o *OnChainBudgetEnforcer) HandleResponse(ctx context.Context, uri string,
+	msg proto.Message) (proto.Message, error) {
+
+	o.Lock()
+	defer o.Unlock()
+
+	checkers := o.checkers()
+	if checkers == nil {
+		return nil, nil
+	}
+
+	checker, ok := checkers[uri]
+	if !ok {
+		return nil, nil
+	}
+
+	if !checker.HandlesResponse(msg.ProtoReflect().Type()) {
+		return nil, fmt.Errorf("invalid implementation, checker for "+
+			"URI %s does not accept response of type %v", uri,
+			msg.ProtoReflect().Type())
+	}
+
+	return checker.HandleResponse(ctx, msg)
+}
+
+// HandleErrorResponse handles and possible alters an error. This is a noop for
+// the OnChainBudgetEnforcer rule.
+//
+// NOTE: this is part of the Enforcer interface.
+func (o *OnChainBudgetEnforcer) HandleErrorResponse(_ context.Context,
+	uri string, respErr error) (error, error) {
+
+	o.Lock()
+	defer o.Unlock()
+
+	checkers := o.checkers()
+	if checkers == nil {
+		return nil, nil
+	}
+
+	checker, ok := checkers[uri]
+	if !ok {
+		return nil, nil
+	}
+
+	return checker.HandleErrorResponse(respErr)
+}
+
+// checkers returns a map of URI to rpcmiddleware.RoundTripChecker which define
+// how the URI should be handled.
+func (o *OnChainBudgetEnforcer) checkers() map[string]mid.RoundTripChecker {
+	// handleOpenChannelRequest is a helper function which handles the
+	// request for opening a single channel.
+	handleOpenChannelRequest := func(ctx context.Context,
+		r *lnrpc.OpenChannelRequest) error {
+
+		if r.FundMax {
+			return fmt.Errorf("max funding not supported")
+		}
+
+		if len(r.Outpoints) > 0 {
+			return fmt.Errorf("outpoints not supported")
+		}
+
+		if r.TargetConf != 0 {
+			return fmt.Errorf("target conf not supported")
+		}
+
+		err := o.checkFeeRate(r.SatPerVbyte)
+		if err != nil {
+			return err
+		}
+
+		amt := uint64(r.LocalFundingAmount + r.PushSat)
+		return o.handlePendingPayment(ctx, &onChainState{Amount: amt})
+	}
+
+	// handleOpenChannelError is a helper function which handles the error
+	// returned for all the channel opening endpoints.
+	handleOpenChannelError := func(_ error) (error, error) {
+		ctx := context.TODO()
+		return nil, o.cancelPendingPayment(ctx)
+	}
+
+	return map[string]mid.RoundTripChecker{
+		"/lnrpc.Lightning/OpenChannelSync": mid.NewFullChecker(
+			&lnrpc.OpenChannelRequest{},
+			&lnrpc.ChannelPoint{},
+			handleOpenChannelRequest,
+			func(ctx context.Context,
+				r *lnrpc.ChannelPoint) (proto.Message, error) {
+
+				return nil, o.handlePaymentConfirmed(ctx)
+			},
+			handleOpenChannelError,
```

Comment:
I'm not sure that this is something we actually can/want to handle, but I thought I'd just point to a potential issue here with using the temp store in between requests and their responses, that could potentially allow on-chain spending above the rule bounds:

Since the temp store is cleared on every restart of litd, we MUST process the response for every successful request in order for it to actually be saved permanently and therefore also be used in the calculation for the current spent on-chain budget after litd is restarted. 

Now, if either:
1. litd is shutdown after a successful request has been sent, but before we've had time to process the response
2. The processing of the response errors (causing a shutdown)

The spent on-chain amount will be lost and will not be used for the future on-chain budget calculations once litd is restarted.

Like I said though, I'm not sure if these edge cases is something we should address, but I just to make you aware of it.

Reply:
Yes, that is (a minor) issue. I'm not sure how that would be solved, because one would need a message buffer on lnd that would persist them in order for litd to be able to ack them? So unless we don't find an easier solution, I'd suggest to keep it this way.

Reply:
Yeah, I agree! Just thought it be worth pointing out.

Reply:
this is a good point and I do think we should consider handling it. In the accounts system, we are "restart" proof so probably worth trying to be restart proof here too if possible. 

for the accounts system, we can do this quite easily cause we know the payment hash to track before sending the request through, so then on startup, we call LND to see if the payment with that hash did in fact go through. 

So is there any way here that we can force there to be some identifier? For example, I see that `OpenChannelRequest` has an optional `Memo string` field and same for `BatchOpenChannel. Can we not force the caller to set these (to something unique that we have not seen before) and then on startup, can we then not list channels & pending channels and then subscribe to channel open events to continue tracking these? For example I see that the Channel open event does contain the memo field. 

If we do this then we also would _not_ use the temp store. Instead we would have some kind of permanent "memo -> amount" map that we can check on startup. 

thoughts?

Reply:
I explored the idea of using the memo to keep track of pending channel opens (see commits after the itest). A missing piece is to add the memo field to the `ClosedChannels` api. With that we could then delete any pending actions for which we don't find an outcome (so that must have either been some internal failure to open the channel or a channel open was abandoned). In my opinion, the complexity is not really worth it. I think we should not use the temporary store, but persist all pending actions and remove them once we get a response/error. This way we are conservative and don't allow overspending. In the worst case this leads to a decreased budget, which may lead to UX issues, but I would consider a session to be something ephemeral and a possible problem can be fixed by just creating a new (non-linked) one.

Reply:
> In my opinion, the complexity is not really worth it. I think we should not use the temporary store, but persist all pending actions and remove them once we get a response/error

agreed. but we should have some identifier persisted with them that is restart safe (ie, not the request ID as used today). We want to make sure that whatever we do now we can potentially recover from in future.  

> but I would consider a session to be something ephemeral and a possible problem can be fixed by just creating a new (non-linked) one.

ah yes that is a good point!